src/libstd/io/mod.rs

```diff
@@ -116,19 +92,45 @@ fn append_to_string<F>(buf: &mut String, f: F) -> Result<usize>
     }
 }
 
+// This uses an adaptive system to extend the vector when it fills. We want to
+// avoid paying to allocate and zero a huge chunk of memory if the reader only
+// has 4 bytes while still making large reads if the reader does have a ton
+// of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
+// time is 4,500 times (!) slower than this if the reader has a very small
+// amount of data to return.
 fn read_to_end<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> Result<usize> {
-    let mut read = 0;
+    let start_len = buf.len();
+    let mut len = start_len;
+    let mut cap_bump = 16;
```

Comment:
Seems like this should be buf.len() rather than a constant. Otherwise, if the vector is large but filled to capacity, you are going to be reallocating it a lot unnecessarily at the beginning of the read.
I think the capacity increase behavior in this function should be identical to what would happen if we used buf.push() on every byte in r (ie it simply doubles the buf's capacity every time it is filled to capacity). This is probably what the user will be expecting, and allows them to be more flexible by reading in data of varying sizes from multiple sources and having it be efficient in every case.